File: 1159-smallest-subsequence-of-distinct-characters/smallest-subsequence-of-distinct-characters.cpp
class Solution {
public:
    string smallestSubsequence(string s) {
        vector<int> last(26);

        // Store last occurrence of every character
        for (int i = 0; i < s.size(); i++)
            last[s[i] - 'a'] = i;

        vector<bool> used(26, false);
        string ans = "";

        for (int i = 0; i < s.size(); i++) {
            char c = s[i];

            // Skip if already included
            if (used[c - 'a'])
                continue;

            // Remove larger characters if they appear later
            while (!ans.empty() &&
                   ans.back() > c &&
                   last[ans.back() - 'a'] > i) {

                used[ans.back() - 'a'] = false;
                ans.pop_back();
            }

            ans.push_back(c);
            used[c - 'a'] = true;
        }

        return ans;
    }
};

File: 1159-smallest-subsequence-of-distinct-characters/smallest-subsequence-of-distinct-characters.py
class Solution:
    def smallestSubsequence(self, s: str) -> str:
        # Dictionary to keep track of the last index of every character
        last_occurrence = {c: i for i, c in enumerate(s)}
        
        stack = []
        seen = set()
        
        for i, c in enumerate(s):
            # If we've already included this character in our result, skip it
            if c in seen:
                continue
            
            # While stack is not empty, AND the current character is smaller than the top of the stack,
            # AND the character at the top of the stack appears again later in the string...
            while stack and c < stack[-1] and last_occurrence[stack[-1]] > i:
                # We can safely discard the top character for now
                removed_char = stack.pop()
                seen.remove(removed_char)
                
            # Add the current character to the stack and mark it as seen
            stack.append(c)
            seen.add(c)
            
        return "".join(stack)